jest.config.packages.ts

```diff
@@ -0,0 +1,209 @@
+
+  // Stop running tests after `n` failures
+  // bail: 0,
+
+  // The directory where Jest should store its cached dependency information
+  // cacheDirectory: "/private/var/folders/fk/c3y07g0576j8_2s9m01pk4qw0000gn/T/jest_dx",
+
+  // Automatically clear mock calls, instances and results before every test.
+  // This does not remove any mock implementation that may have been provided,
+  // so we disable it.
+  // clearMocks: true,
+
+  // Indicates whether the coverage information should be collected while executing the test
+  collectCoverage: true,
+
+  // An array of glob patterns indicating a set of files for which coverage information should be collected
+  collectCoverageFrom: ['./packages/src/**/*.ts'],
+
+  // The directory where Jest should output its coverage files
+  coverageDirectory: 'coverage',
+
+  // An array of regexp pattern strings used to skip coverage collection
+  // TODO: Test index.ts
+  // coveragePathIgnorePatterns: ['./packages/index.ts'],
+
+  // Indicates which provider should be used to instrument code for coverage
+  // XXX: Might have to be 'babel'
+  coverageProvider: 'v8',
+
+  // A list of reporter names that Jest uses when writing coverage reports
+  coverageReporters: ['text', 'html'],
+
+  // An object that configures minimum threshold enforcement for coverage results
+  coverageThreshold: {
+    global: {
+      branches: 90,
+      functions: 96,
+      lines: 95,
+      statements: 95,
+    },
+  },
+
+  // A path to a custom dependency extractor
+  // dependencyExtractor: undefined,
+
+  // Make calling deprecated APIs throw helpful error messages
+  // errorOnDeprecated: false,
+
+  // Force coverage collection from ignored files using an array of glob patterns
+  // forceCoverageMatch: [],
+
+  // A path to a module which exports an async function that is triggered once before all test suites
+  // globalSetup: undefined,
+
+  // A path to a module which exports an async function that is triggered once after all test suites
+  // globalTeardown: undefined,
+
+  // A set of global variables that need to be available in all test environments
+  // globals: {}
+
+  // The maximum amount of workers used to run your tests. Can be specified as % or a number. E.g. maxWorkers: 10% will use 10% of your CPU amount + 1 as the maximum worker number. maxWorkers: 2 will use a maximum of 2 workers.
+  // maxWorkers: "50%",
+
+  // An array of directory names to be searched recursively up from the requiring module's location
+  // moduleDirectories: [
+  //   "node_modules"
+  // ],
+
+  // An array of file extensions your modules use
+  // moduleFileExtensions: [
+  //   "js",
+  //   "jsx",
+  //   "ts",
+  //   "tsx",
+  //   "json",
+  //   "node"
+  // ],
+
+  // A map from regular expressions to module names or to arrays of module names that allow to stub out resources with a single module
+  // NOTE: This must be synchronized with the `paths` option in `tsconfig.packages.json`
+  moduleNameMapper: {
```

Comment:
When Jest runs a test file, it will parse it first and do some processing on the imports in that file. Presumably this is in reference to the option here, in that you can rewrite import paths, but if it can't resolve an import path, then it will fail completely. Here we are telling Jest that when it sees an import path that matches `@metamask/<package-name>`, rewrite it to `packages/<package-name>/src`. This matches the `paths` setting in `tsconfig.packages.js` (see there for more).

Reply:
Yarn 3 creates symlinks in the root `node_modules` for each package, so this shouldn't be necessary? Same for the paths in `tsconfig.json`.

Reply:
I investigated this further. Although it is true that Yarn 3 creates symlinks for `node_modules` for each package, a package has to be built in order to be referenced in a test from another package, because we've configured `package.json` for each package to use `dist/index.js` as the main file. It would be annoying to have to remember to run `yarn build` every time you changed a package that another package was using. What we really want is for Jest to always use the current version of each package, i.e., the code that is located in `src` for each package. In other words, if package `a` imports package `b`, then we don't want `b/dist/index.js` to be loaded, we want `b/src/index.ts` to be loaded instead (and for the TypeScript code to be built on the fly). We can configure TypeScript to do this via `paths` in `tsconfig.json`, but `ts-jest` doesn't have a way to use that option when resolving modules. So this lets us do that from a Jest angle.